kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/SecureConfigInterpolator.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.kubernetes.operator;
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import io.fabric8.kubernetes.api.model.Volume;
+import io.fabric8.kubernetes.api.model.VolumeMount;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * Interpolates a KafkaProxyFilter's configTemplate using a set of {@link SecureConfigProvider}s, producing a complete
+ * configuration for a filter, plus any necessary volumes and volumeMounts for files referred to from that configuration.
+ */
+public class SecureConfigInterpolator {
+
+    private static final Pattern PATTERN = Pattern.compile("^\\$\\{"
+            + "(?<providerName>[a-z]+)"
+            + ":(?<path>[a-zA-Z0-9_.-]+)"
+            + ":(?<key>[a-zA-Z0-9_.-]+)"
+            + "}$");
+
+    private final Map<String, SecureConfigProvider> providers;
+    private final Path mountPathBase;
+
+    public SecureConfigInterpolator(String mountPathBase, Map<String, SecureConfigProvider> providers) {
+        this.providers = providers;
+        this.mountPathBase = Path.of(mountPathBase);
+    }
+
+    InterpolationResult interpolate(Object configTemplate) {
+        // use sets so that it doesn't matter is two providers require the same volume or mount (with exactly the same definition)
+        var volumes = new LinkedHashSet<Volume>();
+        var mounts = new LinkedHashSet<VolumeMount>();
+
+        var interpolated = interpolateRecursive(configTemplate, volumes, mounts);
+
+        return new InterpolationResult(interpolated,
+                volumes, mounts);
+    }
+
+    private @Nullable Object interpolateRecursive(
+                                                  @Nullable final Object jsonValue,
+                                                  @NonNull Set<Volume> volumes,
```

Comment:
Nit: Not a fan of sending down the sets to collect pieces of the ContainerFileReference by side-effect, would rather spend a few objects and have `record Interpolated(Object value, Collection<ContainerFileReference> containerFiles)`. Then build up the sets after the fact.

Here's a PR refactoring this way https://github.com/tombentley/kproxy/pull/18

Reply:
Thanks